client/tufclient.go

```diff
@@ -78,25 +80,68 @@ func (c *TUFClient) update() error {
 	return nil
 }
 
-// downloadRoot is responsible for downloading the root.json
-func (c *TUFClient) downloadRoot() error {
-	role := data.CanonicalRootRole
-	consistentInfo := c.newBuilder.GetConsistentInfo(role)
+// updateRoot checks if there is a newer version of the root available, and if so
+// downloads all intermediate root files to allow proper key rotation.
+func (c *TUFClient) updateRoot() error {
+	// Get Current Root Version
+	currentRootConsistentInfo := c.oldBuilder.GetConsistentInfo(data.CanonicalRootRole)
+	currentVersion := c.oldBuilder.GetLoadedVersion(currentRootConsistentInfo.RoleName)
 
-	// We can't read an exact size for the root metadata without risking getting stuck in the TUF update cycle
-	// since it's possible that downloading timestamp/snapshot metadata may fail due to a signature mismatch
-	if !consistentInfo.ChecksumKnown() {
-		logrus.Debugf("Loading root with no expected checksum")
+	// Get New Root Version
+	raw, err := c.downloadRoot()
+	if _, ok := err.(trustpinning.ErrRootRotationFail); err != nil && !ok {
+		return err
+	}
+	// Ignore rotation errors on the initial download of root
+	err = nil
 
-		// get the cached root, if it exists, just for version checking
-		cachedRoot, _ := c.cache.GetSized(role, -1)
-		// prefer to download a new root
-		_, remoteErr := c.tryLoadRemote(consistentInfo, cachedRoot)
-		return remoteErr
+	signedRoot := &data.Signed{}
+	if err := json.Unmarshal(raw, signedRoot); err != nil {
+		return err
+	}
+	newestRoot, err := data.RootFromSigned(signedRoot)
+	if err != nil {
+		return err
+	}
+	newestVersion := newestRoot.Signed.Version
+	for i := currentVersion; i < newestVersion; i++ {
+		if err := c.updateRootVersions(i, i+1, newestVersion); err != nil {
+			return err
+		}
 	}
+	logrus.Debugf("finished updating root files")
+	return nil
+}
 
-	_, err := c.tryLoadCacheThenRemote(consistentInfo)
-	return err
+// updateRootVersions updates the root from it's current version to a target, rotating keys
+// as they are found
+func (c *TUFClient) updateRootVersions(fromVersion, toVersion, newestVersion int) error {
+	skipCheckExpiryAndChecksum := toVersion != newestVersion
+	logrus.Debugf("updating root from version %d to version %d, up to %d", fromVersion, toVersion, newestVersion)
+
+	var versionedRole string
+	if toVersion == newestVersion {
```

Comment:
This still shouldn't be necessary. Just download the `newestVersion` as `<version>.<role>`. That also mitigates any race conditions where you've pulled and are walking the root rotations and somebody else pushes.